### ISSUE TITLE
LetsEncrypt cannot obtain certificate with RFC2136 Server (HA 0.113.2)

### ISSUE THREAD

**Issue by derekatkins** [home-assistant/addons#1529]
Hi,

I am trying to use LetsEncrypt to get a certificate for a private domain behind a DNS server that I control.  I set up RFC2136 dynamic update but when I configure the LetsEncrypt plugin I get an error.  The log says:

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] file-structure.sh: executing... 
[cont-init.d] file-structure.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[12:05:40] INFO: Selected DNS Provider: dns-rfc2136
[12:05:40] INFO: Use propagation seconds: 60
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator dns-rfc2136, Installer None
Obtaining a new certificate
Performing the following challenges:
dns-01 challenge for name.dom.ain
Cleaning up challenges
Encountered exception during recovery: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/certbot/_internal/auth_handler.py", line 70, in handle_authorizations
    resps = self.auth.perform(achalls)
  File "/usr/local/lib/python3.7/site-packages/certbot/plugins/dns_common.py", line 57, in perform
    self._perform(domain, validation_domain_name, validation)
  File "/usr/local/lib/python3.7/site-packages/certbot_dns_rfc2136/_internal/dns_rfc2136.py", line 76, in _perform
    self._get_rfc2136_client().add_txt_record(validation_name, validation, self.ttl)
  File "/usr/local/lib/python3.7/site-packages/certbot_dns_rfc2136/_internal/dns_rfc2136.py", line 83, in _get_rfc2136_client
    int(self.credentials.conf('port') or self.PORT),
ValueError: invalid literal for int() with base 10: 'null'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/certbot/_internal/error_handler.py", line 125, in _call_registered
    self.funcs[-1]()
  File "/usr/local/lib/python3.7/site-packages/certbot/_internal/auth_handler.py", line 243, in _cleanup_challenges
    self.auth.cleanup(achalls)
  File "/usr/local/lib/python3.7/site-packages/certbot/plugins/dns_common.py", line 76, in cleanup
    self._cleanup(domain, validation_domain_name, validation)
  File "/usr/local/lib/python3.7/site-packages/certbot_dns_rfc2136/_internal/dns_rfc2136.py", line 79, in _cleanup
    self._get_rfc2136_client().del_txt_record(validation_name, validation)
  File "/usr/local/lib/python3.7/site-packages/certbot_dns_rfc2136/_internal/dns_rfc2136.py", line 83, in _get_rfc2136_client
    int(self.credentials.conf('port') or self.PORT),
ValueError: invalid literal for int() with base 10: 'null'
An unexpected error occurred:
ValueError: invalid literal for int() with base 10: 'null'
Please see the logfiles in /var/log/letsencrypt for more details.
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
```
The configuration I am using says:
```
email: me@dom.ain
domains:
  - name.dom.ain
certfile: fullchain.pem
keyfile: privkey.pem
challenge: dns
dns:
  provider: dns-rfc2136
  rfc2136_server: dns.server.name
  rfc2136_name: TheName
  rfc2136_secret: >-
    PART1
    PART2
  rfc2136_algorithm: HMAC-SHA512
```
Do I *NEED* to supply the rfc2136_port setting?  Does it not use the default?
Also, where can I find /var/log/letsencrypt -- and is there a way to run a --dry-run?

**Comment by derekatkins** [comment]
FYI adding "rfc2136_port: 53" to the config gives me a schema error.
If I quote it, tho:  rfc2136_port: '53' -- then it works.  So clearly rfc2136_port is a required parameter and it must be a quoted number.  How... odd.   But at least it's now contacting my DNS server properly.
Either the schema should be changed or the documentation should be updated that you need to quote the port number.
Thanks!

**Comment by ludeeus** [comment]
If you can, please do a PR to update the documentation :+1:

**Comment by derekatkins** [comment]
> If you can, please do a PR to update the documentation +1

Done.  https://github.com/home-assistant/hassio-addons/pull/1532

**Comment by derekatkins** [comment]
Waiting for the PR to be pulled in.

**Comment by derekatkins** [comment]
Still waiting for PR to get pulled.

**Comment by EmnaX** [comment]
Hi @derekatkins,
I have been using the Let's Encrypt add-on successfully with my own DNS servers and the dns-rfc2136 plugin. I have not come across this particular error. I did notice a couple items which may help. You can not use the FQDN for your DNS server (at least I couldn't). I had to use the IP address for it to work successfully. Also, I'm not sure why there is a "Part 1" and "Part 2" for your secret. My secret only had the single long key.

```
email: monkey@hotmail.com
domains:
  - host.example.com
certfile: fullchain.pem
keyfile: privkey.pem
challenge: dns
dns:
  provider: dns-rfc2136
  rfc2136_server: 123.123.123.123
  rfc2136_port: '53'
  rfc2136_name: _acme-challenge.host.example.com. <--whatever your key name is
  rfc2136_zone: example.com
  rfc2136_secret: >-
    hcywJSuvrhjSjksdgeriuhgjbsdkjghWUHhsdhrtiulbsrunarueiHJdfljgfasdffjegdsdlfbirt==
  rfc2136_algorithm: HMAC-SHA512 <-- This MUST be all caps
acme_server: 'https://acme-staging-v02.api.letsencrypt.org/directory' <-- Use this for testing to avoid the rate limits
```

**Comment by derekatkins** [comment]
Hi @EmnaX - 

I've had it working fine since August; I've been waiting for my documentation update to get pulled in.  My issue (and the error above) was because I did not include the `rfc2136_port` setting.  I figured that port 53 was the default, so why would I need to specify the default?  Once I figured out that it required that setting, it all started working for me, which is why I submitted a PR to update the documentation (which is STILL sitting, waiting).

Using FQDN for `rfc2136_server` works for me.  I don't know why it is not working for you.

As for the "Part 1" and "Part 2" in the `secret`, I think it is just one key split into multiple lines.  The "`>-` tag means, IIRC, to ignore whitespace, and my key is pretty long, so the keygen split it up into two parts, 57 chars and 33 chars.

My config looks close to what you posted, although I don't have the `rfc2136_zone` configured, but it's working just fine without it.  And obviously I don't have `acme_server` configured, either, because I have this in production.

It appears they closed the PR without merging, so I guess they don't care that the documentation is broken.

**Comment by EmnaX** [comment]
Oh good to know. I spent a number of hours trying to get this setup working. I edited my comment above with a bit more info about my setup. Either way, hopefully our discussion here will save someone else some time when they try and configure it all!

**Comment by derekatkins** [comment]
It would be nice if the owners would just update the documentation!  That would save other people time and effort, too.   ;-)

**Comment by derekatkins** [comment]
> Oh good to know. I spent a number of hours trying to get this setup working. I edited my comment above with a bit more info about my setup. Either way, hopefully our discussion here will save someone else some time when they try and configure it all!

Thanks.  I just updated and made a new PR.  Care to take a look and see if I missed anything?  I'm happy to update more based on your experiences?

**Comment by derekatkins** [comment]
@EmnaX  --

> I also generated my TSIG key a little differently:
> 
> ```
> $ tsig-keygen -a HMAC-SHA512 _acme-challenge.host.example.com.
> ```
> So my secret was only on a single line without a break in it. 

Indeed.  I recommended a 512-bit key (because why not, when using SHA512), but the default is 256, which is why yours is shorter.

> In my `named.conf.local` file, I have the following line to allow updates for the specific record only:
> 
> ```
>     update-policy {
>         grant _acme-challenge.host.example.com. name _acme-challenge.host.example.com. TXT;
>     };
> ```

Indeed -- I've documented that in the PR already.  Was what I wrote not sufficiently clear?

**Comment by EmnaX** [comment]
@derekatkins ,

> > In my `named.conf.local` file, I have the following line to allow updates for the specific record only:
> > ```
> >     update-policy {
> >         grant _acme-challenge.host.example.com. name _acme-challenge.host.example.com. TXT;
> >     };
> > ```
> 
> Indeed -- I've documented that in the PR already. Was what I wrote not sufficiently clear?

No no, what you have is great! I just included that last blurb about the update-policy just to make my example complete. You have covered it perfectly fine in your PR.

**Comment by derekatkins** [comment]
Still waiting on review/integration or at least some feedback on the documentation.

**Comment by derekatkins** [comment]
Still waiting on review.
